### ISSUE TITLE
When edge weight is changed, MCL result is strange.

### ISSUE THREAD

**Issue by chungfu27** [joandre/MCL_spark#3]
Hi Joandre,

I use your MCL_spark to run MCL in my Spark.
By default (edge weight is 1.0), MCL works fine!
But, when I adjust edge weight, the clustering result is unpredictable.
It means that the all nodes in the same cluster are not connected, for example,
o----o----o (a simple graph)
2----1----2 (the cluster number)
Do you have any suggestion?

**Comment by joandre** [comment]
Hi chungfu27,

Thanks for your message. The last commit seems to correct your issue (at least on the small example you gave me). If it is not ok for your own dataset, please do not hesitate to give it to me so I can check what's going wrong.

Indeed, I am currently in a test phase where I compare my algorithm results with Stijn van Dongen implementation ones and I noticed some differences.

**Comment by chungfu27** [comment]
Hi Joandre,
Thanks for your reply quickly, I attach my dataset, nodes.txt/edges.txt, for you.
[nodes.txt](https://github.com/joandre/MCL_spark/files/75459/nodes.txt)
[edges.txt](https://github.com/joandre/MCL_spark/files/75461/edges.txt)

**Comment by chungfu27** [comment]
Hi Joandre,
I probably find the root cause for this issue.
Because my graph is a weighted graph, the value of self-loop is 1.
If other edge value is very much larger than self-loop value, the probability will be very small through normalization step.
After expansion/inflation, self-loop value will be smaller than epsilon and it will be removed from removeWeakConnections step.
Therefore, the result will be as below:
o----o----o (a simple graph)
2----1----2 (the cluster number)
The middle node will not be in the same cluster.

**Comment by joandre** [comment]
Hi chungfu27,

Thanks for your help. Sorry for the late of my answer. I have decided it was time to build scalaTests. :-)

I think you are right for the reason of this strange behavior. But this not strictly speaking a bug. I have made a quick state-of-the-art and it seems that the best solution is to apply the biggest weight of graph edges to the self-loop for now (called unitary self-loop). In the following paper, there is a very clear explication about the interest (obligation) of having self-loops in graph, in order to avoid non-convergence of MCL.

http://vacher.corinne.free.fr/pdf/Leger_StatsComputing_2013.pdf

To summarize, non-zero weights for self-loops are recommended (for example 1 for weights between 0 and 1), but they can be decreased (for example divide by 10 in the paper) not to "denature" the original graph. In fact, when you add self-loops, you clearly change the graph structure. This a trade-off, that should be up to the user, depending on clusters he obtains and how they make sense in his context. In figure 5, there are different examples of clusters on the same graph, where rk is the inflation rate and ek is the expansion rate. Depending on the strategy you chose, you will obtain different groups.

Perhaps, I should adopt a more open strategy for self-loops.

I will work on that and I will add a clear explanation on the README too, so potential users are aware of this trade-off.
